File: zig/src/fopen.zig
const types = @import("types.zig");
const util = @import("util.zig");
const block_mod = @import("block.zig");
const helpers = @import("fopen_helpers.zig");
const std = @import("std");

pub fn fopen(self: anytype, filename: []const u8, mode: []const u8) ?*types.File {
    if (self.fp == null) {
        return null;
    }
    if (filename.len == 0 or mode.len == 0) {
        return null;
    }

    const bmode: u8 = if (std.mem.eql(u8, mode, "r"))
        0
    else if (std.mem.eql(u8, mode, "w"))
        1
    else if (std.mem.eql(u8, mode, "a"))
        2
    else if (std.mem.eql(u8, mode, "r+"))
        3
    else if (std.mem.eql(u8, mode, "w+"))
        4
    else if (std.mem.eql(u8, mode, "a+"))
        5
    else
        return null;

    var blockindex: u32 = undefined;
    var start: usize = undefined;
    if (filename[0] == '/') {
        blockindex = 1;
        start = 1;
    } else {
        blockindex = util.currentPwdBlockindex(self);
        start = 0;
    }

    var s: [types.BlockNameMaxSize + 2]u8 = [_]u8{0} ** (types.BlockNameMaxSize + 2);
    var slen: usize = 0;
    var i: usize = start;
    while (i < filename.len) : (i += 1) {
        if (filename[i] == '/') {
            if (slen == 0) {
                continue;
            }
            const index = block_mod.findPathBlockindex(self, blockindex, s[0..slen]);
            if (index < 1) {
                return null;
            }
            blockindex = index;
            slen = 0;
            continue;
        }
        s[slen] = filename[i];
        slen += 1;
        if (slen > types.BlockNameMaxSize) {
            return null;
        }
    }
    if (slen == 0) {
        return null;
    }
    const lastname = s[0..slen];
    if (std.mem.eql(u8, lastname, ".") or std.mem.eql(u8, lastname, "..")) {
        return null;
    }

    return switch (bmode) {
        0, 3 => helpers.doFopenR(self, lastname, bmode, blockindex),
        1, 4 => helpers.doFopenW(self, lastname, bmode, blockindex),
        2, 5 => helpers.doFopenA(self, lastname, bmode, blockindex),
        else => null,
    };
}

File: cpp/src/fopen.cpp
#include "types.hpp"

#include "util.hpp"

namespace filefs {

std::optional<File> FileSystem::fopen(const std::string_view filename, const std::string_view mode) {
  if (!state_ || !state_->mounted() || filename.empty() || mode.empty()) {
    return std::nullopt;
  }

  std::uint8_t bmode = 0U;
  if (mode == "r") {
    bmode = 0U;
  } else if (mode == "w") {
    bmode = 1U;
  } else if (mode == "a") {
    bmode = 2U;
  } else if (mode == "r+") {
    bmode = 3U;
  } else if (mode == "w+") {
    bmode = 4U;
  } else if (mode == "a+") {
    bmode = 5U;
  } else {
    return std::nullopt;
  }

  std::uint32_t blockindex = 0U;
  int start = 0;
  if (filename.front() == '/') {
    blockindex = 1U;
    start = 1;
  } else if (state_->tmp.state == 0U) {
    blockindex = state_->pwd_blockindex;
  } else {
    blockindex = state_->tmp.pwd_blockindex;
  }

  std::array<char, BlockNameMaxSize + 2> s{};
  int slen = 0;
  for (int i = start; i < static_cast<int>(filename.size()); ++i) {
    if (filename[static_cast<std::size_t>(i)] == '/') {
      if (slen == 0) {
        continue;
      }
      const auto name = std::string_view{s.data(), static_cast<std::size_t>(slen)};
      const std::uint32_t index = state_->find_path_blockindex(blockindex, name);
      if (index < 1U) {
        return std::nullopt;
      }
      blockindex = index;
      slen = 0;
      continue;
    }
    s[static_cast<std::size_t>(slen++)] = filename[static_cast<std::size_t>(i)];
    if (slen > static_cast<int>(BlockNameMaxSize)) {
      return std::nullopt;
    }
  }
  if (slen == 0) {
    return std::nullopt;
  }

  const std::string_view lastname{s.data(), static_cast<std::size_t>(slen)};
  if (lastname == "." || lastname == "..") {
    return std::nullopt;
  }

  switch (bmode) {
    case 0U:
    case 3U:
      return state_->do_fopen_r(lastname, bmode, blockindex);
    case 1U:
    case 4U:
      return state_->do_fopen_w(lastname, bmode, blockindex);
    case 2U:
    case 5U:
      return state_->do_fopen_a(lastname, bmode, blockindex);
    default:
      return std::nullopt;
  }
}

}  // namespace filefs
